File: cs/src/Connections/ConnectionStatusChangedEventArgs.cs
// <copyright file="ConnectionStatusChangedEventArgs.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license.
// </copyright>

using System;

namespace Microsoft.DevTunnels.Connections;

/// <summary>
/// Connection status change event args.
/// </summary>
public class ConnectionStatusChangedEventArgs : EventArgs
{
    /// <summary>
    /// Create a new instance of <see cref="ConnectionStatusChangedEventArgs"/>.
    /// </summary>
    public ConnectionStatusChangedEventArgs(ConnectionStatus previousStatus, ConnectionStatus status, Exception? disconnectException)
    {
        PreviousStatus = previousStatus;
        Status = status;
        DisconnectException = disconnectException;
    }

    /// <summary>
    /// Get the previous connection status.
    /// </summary>
    public ConnectionStatus PreviousStatus { get; }

    /// <summary>
    /// Get the current connection status.
    /// </summary>
    public ConnectionStatus Status { get; }

    /// <summary>
    /// Get the exception that caused disconnect if <see cref="Status"/> is <see cref="ConnectionStatus.Disconnected"/> and the tunnel client is not disposed yet.
    /// <c>null</c> if <see cref="Status"/> is not <see cref="ConnectionStatus.Disconnected"/> or the tunnel client is disposed.
    /// </summary>
    /// <seealso cref="TunnelConnection.DisconnectException"/>.
    public Exception? DisconnectException { get; }
}

File: ts/src/connections/connectionStatusChangedEventArgs.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { ConnectionStatus } from './connectionStatus';

/**
 * Connection status changed event args.
 */
export class ConnectionStatusChangedEventArgs {
    /**
     * Creates a new instance of ConnectionStatusChangedEventArgs.
     */
    public constructor(
        /**
         * Gets the previous connection status.
         */
        public readonly previousStatus: ConnectionStatus,

        /**
         * Gets the current connection status.
         */
        public readonly status: ConnectionStatus,

        /**
         * Gets the error that caused disconnect if {@link status} is {@link ConnectionStatus.Disconnected}.
         */
        public readonly disconnectError?: Error,
    ) {}
}
